gnucash/import-export/test/gtest-import-backend.cpp

```diff
@@ -0,0 +1,258 @@
+        xaccTransCommitEdit(trans);
+
+        g_date_free(date);
+
+        return trans;
+    }
+
+    MockPrefsBackend *m_prefs;
+    QofBook *m_book {};
+    Account *m_import_acc {};
+};
+
+
+TEST_F(ImportBackendTest, CreateTransInfo)
+{
+    using namespace testing;
+
+    Account *dest_acc1 = CreateAccount("Destination 1");
+    Account *dest_acc2 = CreateAccount("Destination 2");
+    Transaction *trans = CreateTransaction("This is the description", 1.0);
+    MockGncImportMatchMap imap;
+
+    EXPECT_CALL(imap, findAccount(_, StrEq("This is the description")))
+        .Times(1)
+        .WillOnce(Return(dest_acc1));
+    GNCImportTransInfo *trans_info = gnc_import_TransInfo_new(trans, &imap);
+    EXPECT_EQ(gnc_import_TransInfo_get_destacc(trans_info), dest_acc1);
+
+    gnc_import_TransInfo_set_destacc(trans_info, dest_acc2, false);
+    EXPECT_EQ(gnc_import_TransInfo_get_destacc(trans_info), dest_acc2);
+
+    gnc_import_TransInfo_delete(trans_info);
+};
+
+
+class ImportBackendBayesTest : public ImportBackendTest
+{
+protected:
+    void SetUp()
+    {
+        ImportBackendTest::SetUp();
+
+        using namespace testing;
+
+        // set bayesian import matching in preferences
+        EXPECT_CALL(*m_prefs, getBool(StrEq(GNC_PREFS_GROUP_IMPORT), StrEq(GNC_PREF_USE_BAYES)))
+            .Times(AnyNumber())
+            .WillRepeatedly(Return(true));
+    }
+};
+
+
+TEST_F(ImportBackendBayesTest, CreateTransInfo)
+{
+    using namespace testing;
+
+    Account *dest_acc = CreateAccount("Destination");
+    MockGncImportMatchMap imap;
+
+    // This transaction is used for testing tokenization of its content.
+    // Therefore the description text and the memo should contain
+    //   * consecutive separators
+    //   * separators at the beginning and end of string
+    //   * duplicated tokens within and between description text end memo
+    // The token separator is space.
+    Transaction *trans = CreateTransaction(" test  tokens within   description  tokens  ", 1.0, "  test   the memo test ");
+
+    // check tokens created from transaction
+    EXPECT_CALL(imap, findAccountBayes(AllOf(
+            Each(Not(IsEmpty())),             // tokens must not be empty strings
+            Each(Not(HasSubstr(" "))),        // tokens must not contain separator
+            Not(HasDuplicates()),             // tokens must be unique
+            Contains(StrEq("description")),   // spot sample
+            Contains(StrEq("memo"))           // spot sample
+            )))
+        .Times(1)
+        .WillOnce(Return(dest_acc));
+    GNCImportTransInfo *trans_info = gnc_import_TransInfo_new(trans, &imap);
+    EXPECT_EQ(gnc_import_TransInfo_get_destacc(trans_info), dest_acc);
+
+    gnc_import_TransInfo_delete(trans_info);
+};
+
+
+// main test function
+int main(int argc, char **argv)
+{
+    testing::InitGoogleTest(&argc, argv);
+
+    qof_init();
+    if (!cashobjects_register())
+        exit(1);
+
+    return RUN_ALL_TESTS();
+}
```

Comment:
You don't need a main(). Googletest takes care of that.

Reply:
That's correct if only `testing::InitGoogleTest()` and `RUN_ALL_TESTS()` is called. Here I additionally wanted to initialize QOF and call `cashobjects_register()`.

When I'll mock `libgncmod-engine` completely now, this will not be necessary anymore, I guess, and `main()` function can be omitted.

Reply:
You're building it with `gnc_add_test()` in CMakeLists.txt. That will set up the program for you and call those functions. Having your own `main` should fail to compile because you're not allowed to have two.
GoogleTest provides a [Global SetUp and TearDown](https://github.com/google/googletest/blob/master/googletest/docs/advanced.md#global-set-up-and-tear-down) facility where you can do things like call `qof_init` and `cashobjects_register`. *But* calling either of those gets you back to linking libgncmod-engine.

Reply:
Ok, good to know.